README.md
# lua-native-ssi-nginx

* Latest Release: [![GitHub version](https://badge.fury.io/gh/DracoBlue%2Flua-native-ssi-nginx.png)](https://github.com/DracoBlue/lua-native-ssi-nginx/releases)
* Build Status: [![Build Status](https://github.com/dracoblue/lua-native-ssi-nginx/actions/workflows/run-tests.yml/badge.svg?branch=master)](https://github.com/DracoBlue/lua-native-ssi-nginx/actions/workflows/run-tests.yml)

This is an effort to replace nginx's c ssi implementation with a flexible native lua based version, since nginx ssi does
[not](https://github.com/openresty/lua-nginx-module#mixing-with-ssi-not-supported) work with the lua module.

This solution has some advantages over the c ssi version:

* it allows regexp for ssi types (because there are [no wildcards](http://stackoverflow.com/questions/34392175/using-gzip-types-ssi-types-in-nginx-with-wildcard-media-types) in c ssi_types)
* it works with lua module
* for `200 OK` responses it generates and handles etags based on md5 *after* all ssi includes have been performed
* it handles and sanitizes invalid json in subrequests (inline or as summary)
* it handles **only**: `<!--#include file="PATH" -->` and `<!--#include virtual="PATH" -->` and no other ssi features
* it minimizes `max-age` of `Cache-Control` to the lowest value 

## Usage

If you started with a location like this:

``` txt
location / {
	proxy_pass http://127.0.0.1:4777;
	# add your proxy_* parameters and so on here
}
```

you have to replace it with something like this:

``` txt
location /ssi-api-gateway/ {
	internal;
	rewrite ^/ssi-api-gateway/(.*)$ /$1  break;
	proxy_pass http://127.0.0.1:4777;
	# add your proxy_* parameters and so on here
}

location / {
	set $ssi_api_gateway_prefix "/ssi-api-gateway";
	set $ssi_validate_json_types "application/json application/.*json";
	set $ssi_invalid_json_fallback '{"error": "invalid json", "url": %%URL%%, "message": %%MESSAGE%%}';
	content_by_lua_file "/etc/nginx/lua-ssi-content.lua";
	header_filter_by_lua_file "/etc/nginx/lua-ssi-header.lua";
}
```

The `ssi-api-gateway` location is necessary to use e.g. nginx's caching layer and such things.

## Activate SSI only for specific content types

If you want to enable ssi only for specific content types, use the following nginx configuration variable in the nginx
location:

``` txt
set $ssi_types "text/.*html application/.*json";
```

The default is:

``` txt
set $ssi_types ".*";
```


## Activate JSON Summary Validation

**Prerequisites**: Install cjson (e.g. `apt-get install lua-cjson` to activate this feature. Otherwise you get the following message:
`Even though ssi_validate_json is true, the cjson library is not installed! Skip validation!`.

If you want to ensure, that subrequested json is always valid, you can activate this in the nginx location:

``` txt
set $ssi_validate_json_types "application/json application/.*json";
set $ssi_invalid_json_fallback '{"error": "invalid json", "url": %%URL%%, "message": %%MESSAGE%%}';
```

If you setup the configuration like this, the following ssi:

``` txt
GET /broken_json_include/
{"thisIsThe": "index", "sub_resources": [<!--# include file="/broken_json_include/broken_sub_resource.json" -->] }

GET /broken_json_include/broken_sub_resource.json
{"thisIsA": "subResource", "with invalud json}
```


will result in the following valid json response:

``` json
{
	"error": "invalid json",
	"brokenSsiRequests": [
		{
			"url": "\/broken_json_include\/broken_sub_resource.json",
			"message": "Expected object key string but found unexpected end of string at character 47"
		}
	],
	"message": "Expected object key string but found unexpected end of string at character 91","url":"\/broken_json_include\/"
}
```

## Activate JSON Inline Validation

If you don't want to replace the entire SSI response with an error summary (like in the previous section), you can add:

```
set $ssi_validate_json_inline on;
```

and only the broken SSI will be replaced with the `$ssi_invalid_json_fallback`.

**Important**: Please don't forget to define `$ssi_validate_json_types` and `$ssi_invalid_json_fallback` like described
in the previous section!.

So:

``` txt
GET /broken_json_include/
{"thisIsThe": "index", "sub_resources": [<!--# include file="/broken_json_include/broken_sub_resource.json" -->] }

GET /broken_json_include/broken_sub_resource.json
{"thisIsA": "subResource", "with invalud json}
```

will result in the following valid json response:

``` json
{
  "thisIsThe": "index",
  "sub_resources": [
    {
      "error": "invalid json",
      "url": "/broken_json_include/",
      "message": "Expected object key string but found unexpected end of string at character 47"
    }
  ]
}
```

## Limit recursion depth

The default values for the maximum depth (1024) and the maximum amount of includes (65535) can be changed with the following
configuration parameters:

```
set $ssi_max_includes 512;
set $ssi_max_ssi_depth 16;
```

If the limit is exceeded, the ssi will be replaced with:

``` json
{
	"error": "invalid json",
	"url": "\/recursion_cap_depth\/sub_resource.json",
	"message": "max recursion depth exceeded 16(was 17)"
}
```

or

``` json
{
	"error": "invalid json",
	"url": "\/recursion_cap\/sub_resource.json",
	"message": "max ssi includes exceeded 512(was 728)"
}
```


You can change the response with:

``` json
set $ssi_invalid_json_fallback '{"error": "invalid json", "url": %%URL%%, "message": %%MESSAGE%%}';
```

## Minimize `max-age` in `Cache-Control`

You can calculate the lowest `max-age` of the root document and all sub resources and return the lowest value. Additionally
 it takes the `age` response header of the sub resources into account and decreases the `max-age` by this value. This
 feature is opt-in only and you can activate it like this:

``` txt
set $ssi_minimize_max_age on;
```

The default is:

``` txt
set $ssi_minimize_max_age off;
```

An example:

    /users (max-age=60, age=0 -> ttl=60), includes:
       -> /users/1 (max-age=10, age=7 -> ttl=3)
       -> /users/2 (max-age=5, age=0 -> ttl=5)

will return in `max-age=3` since 3 is the lowest ttl and thus the `max-age` value for the entire request.

**Important**: If you activate this feature, all other Cache-Control directives will be removed and only `Cache-Control: max-age=300`
(if the minimum max-age was 300) or `Cache-Control: max-age=0, no-cache` (if the minimum was negative) will be served.
 Additional Cache-Control features like `stale-while-revalidate` or `stale-if-error` will be removed.
 
Invalid max-age values will be replaced with `Cache-Control: no-cache, max-age=0`.

Additonally you may use:

```
		set $ssi_minimize_override_stale_while_revalidate 5;
```

to append `stale-while-revalidate=5` to each `Cache-Control` header with `max-age` greater than 0.

Starting with Version 1.7.0: If no `Cache-Control` header is send by subrequest, it will be handled like `max-age=0`. There is a header called `X-Ssi-Missing-CC-Count: 2`,
which makes the amount of subrequests missing a cache control header in this request. It will also appear in the log like this: 

```
2022/01/30 09:17:05 [error] 7#7: *487 [lua] lua-ssi-content.lua:301: missing cache control on sub request url: /ssi-api-gateway/max-age/no-cache-control.json while sending to client, client: 172.17.0.1, server: , request: "GET /max-age/include-without-cache-control-expires-in-30.json HTTP/1.1", host: "localhost:4778"
```

and on root it looks like this:

```
2022/01/30 09:17:05 [error] 7#7: *487 [lua] lua-ssi-content.lua:301: missing cache control on root request url: /ssi-api-gateway/max-age/no-cache-control.json while sending to client, client: 172.17.0.1, server: , request: "GET /max-age/no-cache-control.json HTTP/1.1", host: "localhost:4778"
```


## Debug Output

If you want to debug the lua ssi output or calculations, you should enable the debug log in nginx.

If you need to debug one request and don't want to enable debug log for the entire server, you can send a special request
header called `X-Ssi-Debug: true`.

    $ curl -v -sS -H 'X-Ssi-Debug: true' "http://localhost:4778/max-age/include-stale-expires-in-120.json" 2>&1
    < X-Ssi-Minimize-MaxAge-Url: /max-age/35-seconds/30-age/40-swr
    < X-Ssi-Minimize-MaxAge-Age: 45
    < X-Ssi-Minimize-MaxAge-Cache-Control: max-age=35, stale-while-revalidate=40
    
If you apply the `X-Ssi-Debug: true` request header, you will receive extra `X-Ssi-Minimize`-prefixed response headers.
In those headers you can see, which of the ssi requests was the one, which resulted in the final max-age of the
`Cache-Control` response header.

## Development

To run the tests locally launch:

``` console
$ ./launch-test-nginx.sh
...
Successfully built 72a844684987
2016/09/11 11:34:02 [alert] 1#0: lua_code_cache is off; this will hurt performance in /etc/nginx/sites-enabled/port-4778-app.conf:12
nginx: [alert] lua_code_cache is off; this will hurt performance in /etc/nginx/sites-enabled/port-4778-app.conf:12
```

Now the nginx processes are running with docker.

Now you can run the tests like this:

``` console
$ ./run-tests.sh
  [OK] echo
  [OK] echo_custom_header
  [OK] echo_method
  [OK] gzip
  [OK] image
  [OK] json
  [OK] json_include
  [OK] one
```

## FAQ

### Subrequests hang when used with lua native ssi

The following setup is often used, to avoid buffering on disk:

``` text
proxy_buffer_size          16k;
proxy_buffering         on;
proxy_max_temp_file_size 0;
```

but it will result in hanging requests, if the response size is bigger then 16k.

That's why you should either use (means: disable buffering at all):

``` text
proxy_buffer_size          16k;
proxy_buffering         off;
```

or (means: store up to 1024m in temp file)

``` text
proxy_buffer_size          16k;
proxy_buffering         on;
proxy_max_temp_file_size 1024m;
```

to work around this issue.


## TODOs

See <https://github.com/DracoBlue/lua-native-ssi-nginx/issues> for all open TODOs.

## Changelog

See [CHANGELOG.md](./CHANGELOG.md).

## License

This work is copyright by DracoBlue (<http://dracoblue.net>) and licensed under the terms of MIT License.


CHANGELOG.md
# lua-native-ssi-nginx CHANGELOG

## 1.7.0

* Added Header `X-Ssi-Missing-CC-Count` to count the amount of missing cache control headers
* Added log if cache-control was missing but cache minimizing is activated

## 1.6.0

- added `X-Ssi-Debug: true` request header to debug max age minimize behaviour
- fixed `nocache` to `no-cache` in `cache-control` header

## 1.5.0

- added config to override stale-while-revalidate for minimized cache
- handle stale-while-revalidate when minimizing max-cache
- log subrequest url in debug log

## 1.4.2

- set age to 0 if cache control max age is minizimed

## 1.4.1

- allow spaces around cache control directives

## 1.4.0

- added possibility to minimize `max-age` of the response by `age` and `max-age` of all sub requests

## 1.3.0

- render ssi error, if relative path is in ssi include

## 1.2.0

- fixed internal server error on percent in url or message

## 1.1.0

- added explanation for `proxy_max_temp_file_size 0` vs `proxy_buffering on`
- removed `always_forward_body` because it does not always send all data to the first request

## 1.0.3

- fixed url in inline json validation response

## 1.0.2

- added recursion handling (depth and max includes)

## 1.0.1

- handle header-only invocations (don't crash on empty `ngx.ctx.res`)
- removed necessity for `lua_need_request_body on` in nginx
- forward request body in a native way now (with `always_forward_body`)

## 1.0.0

- initial release


run-tests.sh
#!/usr/bin/env bash

test_success=0
test_errors=0

cd `dirname $0`
cd tests

# We cannot use while read line here
# @see http://fvue.nl/wiki/Bash:_Piped_%60while-read'_loop_starts_subshell
# for further information
for file in `ls *.txt`
do
	TEST_NAME=`echo "$file" | cut -f 1 -d '.'`
	if [ -f "$TEST_NAME.sh" ]
	then
		bash "$TEST_NAME.sh" $1 > "${TEST_NAME}.result"
		current_exit_code="${?}"
	elif [ -f "$TEST_NAME.lua" ]
	then
		docker run -it -v `pwd`/../:/usr/src/app --workdir /usr/src/app --rm pirogoeth/alpine-lua:5.2 lua5.2 tests/$TEST_NAME.lua > "${TEST_NAME}.result"
		current_exit_code=0
	else
		curl -sS "localhost:4778/${TEST_NAME}/" > "${TEST_NAME}.result"
		current_exit_code="${?}"
	fi
	if [ "${current_exit_code}" -ne "0" ]
	then
		echo "  [  ] $TEST_NAME"
		echo "   -> broken! (curl did not 2xx, Exit code: $current_exit_code)"
		let test_errors=test_errors+1
	else
		diff "${TEST_NAME}.txt" "${TEST_NAME}.result"
		current_exit_code="${?}"
		if [ "${current_exit_code}" -ne "0" ]
		then
			echo "  [  ] $TEST_NAME"
			echo "   -> broken! (.txt != .result, Exit code: $current_exit_code)"
			let test_errors=test_errors+1
		else
			echo "  [OK] $TEST_NAME"
			let test_success=test_success+1
		fi
	fi
done

if [ ! $test_errors -eq 0 ]
then
	exit 1
fi

exit 0


tests/echo.sh
#!/usr/bin/env bash

curl -sS -X POST -H "Content-Type: application/json" -d '{"key":"value"}' "localhost:4778/echo/"
exit $?

tests/etag_check.sh
#!/usr/bin/env bash

curl -v -sS -X POST -H "Content-Type: application/json" -d '{"key":"value"}' "localhost:4778/echo/" 2>&1 | grep '< ETag' | tr -d "\n"
exit $?

src/lua-ssi-header.lua
if not ngx.headers_sent and ngx.ctx.res
then
    local requestHeaders = {}

    local sanatizeHeaderFieldName = function(headerFieldName)
        return string.gsub(string.lower(headerFieldName), "_", "-")
    end

    for k, v in pairs(ngx.req.get_headers()) do
        k = sanatizeHeaderFieldName(k)
        requestHeaders[k] = v
    end

    for k, v in pairs(ngx.ctx.res.header) do
        ngx.header[k] = v
    end
    if ngx.ctx.ssiRequestsCount then
        ngx.header["X-Ssi-Sub-Requests"] = ngx.ctx.ssiRequestsCount
    end
    if ngx.ctx.ssiIncludesCount then
        ngx.header["X-Ssi-Includes"] = ngx.ctx.ssiIncludesCount
    end
    if ngx.ctx.ssiDepth then
        ngx.header["X-Ssi-Depth"] = ngx.ctx.ssiDepth
    end
    if ngx.ctx.ssiMissingCacheControlCount then
        ngx.header["X-Ssi-Missing-CC-Count"] = ngx.ctx.ssiMissingCacheControlCount
    end
    if requestHeaders["x-ssi-debug"] == "true" then
        ngx.header["X-Ssi-Minimize-MaxAge-Url"] = ngx.ctx.ssiMinimizeMaxAgeUrl
        ngx.header["X-Ssi-Minimize-MaxAge-Age"] = ngx.ctx.ssiMinimizeMaxAgeAge
        ngx.header["X-Ssi-Minimize-MaxAge-Cache-Control"] = ngx.ctx.ssiMinimizeMaxAgeCacheControl
    end
    if ngx.ctx.overrideContentType then
        ngx.header["Content-Type"] = ngx.ctx.overrideContentType
    end
    if ngx.ctx.overrideCacheControl then
        ngx.header["Cache-Control"] = ngx.ctx.overrideCacheControl
        ngx.header["Age"] = "0"
    end
    ngx.header["Content-Length"] = nil
    if ngx.ctx.etag then
        ngx.header["ETag"] = ngx.ctx.etag
        local ifNoneMatch = ngx.req.get_headers()["If-None-Match"] or nil
        ngx.log(ngx.DEBUG, "If-None-Match: ", ifNoneMatch)
        ngx.log(ngx.DEBUG, "ETag: ", ngx.ctx.etag)

        if ifNoneMatch == ngx.ctx.etag
        then
            ngx.header["Content-Length"] = 0
            ngx.exit(ngx.HTTP_NOT_MODIFIED)
            return
        end
    end
end


src/lua-ssi-content.lua

local prefix = ngx.var.ssi_api_gateway_prefix
local invalidJsonFallback = ngx.var.ssi_invalid_json_fallback or '{"error": "invalid json in ssi", "url": %%URL%%, "message": %%MESSAGE%%}'
local maxSsiDepth = 1024
local maxSsiIncludes = 65535
if ngx.var.ssi_max_includes ~= nil and ngx.var.ssi_max_includes ~= ""
then
    maxSsiIncludes = tonumber(ngx.var.ssi_max_includes)
end
if ngx.var.ssi_max_ssi_depth ~= nil and ngx.var.ssi_max_ssi_depth ~= ""
then
    maxSsiDepth = tonumber(ngx.var.ssi_max_ssi_depth)
end

local validateJson = false
local validateJsonInline = false
local validateJsonTypes = {}
if ngx.var.ssi_validate_json_types ~= nil and ngx.var.ssi_validate_json_types ~= ""
then
    validateJson = true
    validateJsonTypes = string.gmatch(ngx.var.ssi_validate_json_types, "%S+")
end
if ngx.var.ssi_validate_json_inline ~= nil and ngx.var.ssi_validate_json_inline == "on"
then
    validateJsonInline = true
end
local ssiTypes = string.gmatch(".*", "%S+")
if ngx.var.ssi_types ~= nil and ngx.var.ssi_types ~= ""
then
    ssiTypes = string.gmatch(ngx.var.ssi_types, "%S+")
end
local minimizeMaxAge = false
if ngx.var.ssi_minimize_max_age ~= nil and ngx.var.ssi_minimize_max_age == "on"
then
    minimizeMaxAge = true
end
local minimizeOverrideStaleWhileRevalidate = false
if ngx.var.ssi_minimize_override_stale_while_revalidate ~= nil and ngx.var.ssi_minimize_override_stale_while_revalidate ~= ""
then
    minimizeOverrideStaleWhileRevalidate = tonumber(ngx.var.ssi_minimize_override_stale_while_revalidate)
end

local sanatizeHeaderFieldName = function(headerFieldName)
    return string.gsub(string.lower(headerFieldName), "_", "-")
end

ngx.req.read_body()

local res = ngx.location.capture(
    prefix .. ngx.var.request_uri, {
        method = ngx["HTTP_" .. ngx.var.request_method],
        body = ngx.req.get_body_data()
    }
)

getSanitizedFieldFromHeaders = function(rawFieldName, headers)
    local sanatizedFieldName = sanatizeHeaderFieldName(rawFieldName)
    for k, v in pairs(headers) do
        if sanatizeHeaderFieldName(k) == sanatizedFieldName
        then
            return v
        end
    end

    return nil
end

getCacheControlFieldsFromHeaders = function(headers)
    local cacheControlHeader = getSanitizedFieldFromHeaders("cache-control", headers)
    if not cacheControlHeader then
        return {}
    end

    local cacheControlHeaderPrefixedAndSuffixedWithAWhitespace = ", " .. cacheControlHeader .. " ,"

    local fields = {}
    
    for key in string.gmatch(cacheControlHeaderPrefixedAndSuffixedWithAWhitespace, ',[%s]-([^=%s,]+)[%s]-')
    do
        fields[key] = true
    end

    for key, value in string.gmatch(cacheControlHeaderPrefixedAndSuffixedWithAWhitespace, '[%s,]+([^=%s,]+)%s-=%s-([^%s,]+)[%s,]-')
    do
        fields[key] = value
    end

    for key, value in string.gmatch(cacheControlHeaderPrefixedAndSuffixedWithAWhitespace, '[%s,]+([^=%s,]+)%s-=%s-"([^"]+)"[%s,]-')
    do
        fields[key] = value
    end

    return fields
end

getMaxAgeDecreasedByAgeOrZeroFromHeaders = function(headers)
    local respCacheControlFields = getCacheControlFieldsFromHeaders(headers)
    local respCacheControlMaxAge = (respCacheControlFields["max-age"] ~= nil and tonumber(respCacheControlFields["max-age"])) or nil
    if respCacheControlMaxAge == nil
    then
        if respCacheControlFields["max-age"] ~= nil then
            ngx.log(ngx.ERR, "request cache-control max-age is an invalid number: " .. tostring(respCacheControlFields["max-age"]))
        end
        return 0, tonumber(respCacheControlFields["stale-while-revalidate"] or 0)
    end

    local respCacheAge = tonumber(getSanitizedFieldFromHeaders("age", headers));    
    local respCacheSwr = tonumber(respCacheControlFields["stale-while-revalidate"]);
    
    ngx.log(ngx.DEBUG, "request cache-control: " .. tostring(respCacheControlMaxAge) .. " and age: " .. tostring(respCacheAge) .. " and stale-while-revalidate: " .. tostring(respCacheSwr));

    if respCacheAge ~= nil then
        respCacheControlMaxAge = respCacheControlMaxAge - respCacheAge
    end
    
    if respCacheSwr ~= nil then
        respCacheControlMaxAge = respCacheControlMaxAge + respCacheSwr
    else
        if respCacheControlFields["stale-while-revalidate"] ~= nil then
            ngx.log(ngx.ERR, "request cache-control stale-while-revalidate is an invalid number: " .. tostring(respCacheControlFields["stale-while-revalidate"]))
        end
    end

    if respCacheControlMaxAge < 0 then
        respCacheControlMaxAge = 0
    end

    return respCacheControlMaxAge, tonumber(respCacheControlFields["stale-while-revalidate"] or 0)
end

getContentTypeFromHeaders = function(headers)
    return getSanitizedFieldFromHeaders("content-type", headers)
end

matchesContentTypesList = function(contentType, contentTypesList)
    if contentType == nil
    then
        return false
    end

    for contentTypeListItem in contentTypesList do
        if string.match(contentType, contentTypeListItem)
        then
            return true
        end
    end
    return false
end

local cjson = (function(validateJson)
    if not validateJson then
        return false
    end

    local hasCjson, cjson = pcall(function()
        return require "cjson.safe"
    end)

    if (hasCjson)
    then
        return cjson
    end

    return false
end)(validateJson)

local generateJsonErrorFallback = function(url, message)
    local escapedUrl = cjson.encode(url)
    local escapedMessage = cjson.encode(message)

    local body = string.gsub(invalidJsonFallback, "%%%%URL%%%%", function()
        return escapedUrl
    end)
    return string.gsub(body, "%%%%MESSAGE%%%%", function()
        return escapedMessage
    end)
end

if validateJson and not cjson then
    ngx.log(ngx.ERR, "Even though ssi_validate_json is true, the cjson library is not installed! Skip validation!")
end

ngx.log(ngx.DEBUG, "request_uri: ", prefix .. ngx.var.request_uri)
local captureRegularFileExpression = '<!%-%-#%s*include file="([^"]+)"%s*%-%->'
local captureRegularVirtualExpression = '<!%-%-#%s*include virtual="([^"]+)"%s*%-%->'
local captureRegularFileExpressions = {captureRegularFileExpression,captureRegularVirtualExpression}

local getSsiRequestsAndCount = function(ssiResponses, body)
    local ssiRequests = {}
    local ssiRequestsCount = 0
    local ssiMatchesCount = 0
    local ssiRequestLock = {}

    for i,captureRegularExpression in ipairs(captureRegularFileExpressions) do
        local regularExpression = string.gsub(captureRegularExpression, "([%(%)])", "")
        local matches = string.gmatch(body, regularExpression)
        for match,n in matches do
--          ngx.log(ngx.DEBUG, "matches", match)
            local ssiVirtualPath = string.match(match, captureRegularExpression)
--          ngx.log(ngx.DEBUG, "ssiVirtualPath", ssiVirtualPath)
            if ssiResponses[prefix .. ssiVirtualPath] == nil and ssiRequestLock[prefix .. ssiVirtualPath] == nil
--          if ssiResponses[prefix .. ssiVirtualPath] == nil
            then
                ssiRequestLock[prefix .. ssiVirtualPath] = true
                if string.sub(ssiVirtualPath, 0, 1) == "/" then
                    table.insert(ssiRequests, { prefix .. ssiVirtualPath })
                    ssiRequestsCount = ssiRequestsCount + 1
                else
                    ssiResponses[prefix .. ssiVirtualPath] = {status = 500, header = {}, body = generateJsonErrorFallback(ssiVirtualPath, "ssi virtual path must start with a /")}
                end
            end
            ssiMatchesCount = ssiMatchesCount + 1
        end
    end

    return ssiRequests, ssiRequestsCount, ssiMatchesCount
end

if res then
    local contentType = getContentTypeFromHeaders(res.header)
    ngx.status = res.status
--    ngx.say("status: ", res.status)
--    ngx.say("body:")
--    ngx.print(res.body)
    local body = res.body
    local minimumCacheControlMaxAge = nil
    local rootCacheControlMaxAge = nil
    local rootCacheControlSwr = nil
    local totalMissingCacheControlCount = 0
    if minimizeMaxAge then
        rootCacheControlMaxAge, rootCacheControlSwr = getMaxAgeDecreasedByAgeOrZeroFromHeaders(res.header)
        if (getSanitizedFieldFromHeaders("cache-control", res.header) == nil) then
            ngx.log(ngx.ERR, "missing cache control on root request url: " .. ngx.var.request_uri)
            totalMissingCacheControlCount = totalMissingCacheControlCount + 1
        end
        ngx.ctx.ssiMinimizeMaxAgeUrl = ngx.var.request_uri
        ngx.ctx.ssiMinimizeMaxAgeAge = rootCacheControlMaxAge
        ngx.ctx.ssiMinimizeMaxAgeCacheControl = getSanitizedFieldFromHeaders("cache-control", res.header)
        minimumCacheControlMaxAge = rootCacheControlMaxAge
        if rootCacheControlMaxAge == 0 then
            rootCacheControlMaxAge = nil
        end
        ngx.log(ngx.DEBUG, "cache-control root: " .. tostring(rootCacheControlMaxAge))
    end

    if (validateJson)
    then
        ngx.log(ngx.DEBUG, "check if content type matches: ", contentType)
        validateJson = matchesContentTypesList(contentType, validateJsonTypes)
    end

    if matchesContentTypesList(contentType, ssiTypes)
    then
        local ssiResponses = {}
        local totalSsiSubRequestsCount = 0
        local totalSsiIncludesCount = 0
        local totalSsiDepth = 0

        local ssiRequests, ssiRequestsCount, ssiMatchesCount = getSsiRequestsAndCount(ssiResponses, body)

        while ssiMatchesCount > 0
        do
            totalSsiDepth = totalSsiDepth + 1
            if (totalSsiDepth > maxSsiDepth or totalSsiIncludesCount > maxSsiIncludes) and ssiMatchesCount > 0
            then
                if (totalSsiDepth > maxSsiDepth)
                then
                    ngx.log(ngx.ERR, "max recursion depth exceeded " .. maxSsiDepth .. "(was " .. totalSsiDepth .. ")")
                else
                    ngx.log(ngx.ERR, "max ssi includes exceeded " .. maxSsiIncludes .. "(was " .. totalSsiIncludesCount .. ")")

                end
                for i,captureRegularExpression in ipairs(captureRegularFileExpressions) do
                    local regularExpression = string.gsub(captureRegularExpression, "([%(%)])", "")

                    local replacer = function(w)
                        local ssiVirtualPath = string.match(w, captureRegularExpression)
                        if (totalSsiDepth > maxSsiDepth)
                        then
                            return generateJsonErrorFallback(ssiVirtualPath, "max recursion depth exceeded " .. maxSsiDepth .. "(was " .. totalSsiDepth .. ")")
                        else
                            return generateJsonErrorFallback(ssiVirtualPath, "max ssi includes exceeded " .. maxSsiIncludes .. "(was " .. totalSsiIncludesCount .. ")")
                        end
                    end

                    body = string.gsub(body, regularExpression, replacer)
                end

                ssiMatchesCount = 0
            else
                if (ssiRequestsCount > 0)
                then
                    -- FIXME: handle ssiRequestsCount > 200, because this is the internal nginx limit
                    -- issue all the requests at once and wait until they all return
                    local resps = { ngx.location.capture_multi(ssiRequests) }

                    -- loop over the responses table
                    for i, resp in ipairs(resps) do
                        --            ngx.log(ngx.DEBUG, "resp ", i, " with ", resp.status, " and body ", resp.body)
                        ngx.log(ngx.DEBUG, "sub request url ", ssiRequests[i][1], " and status ", resp.status)
                        if minimizeMaxAge and minimumCacheControlMaxAge ~= nil then
                            local respCacheControlMaxAge = getMaxAgeDecreasedByAgeOrZeroFromHeaders(resp.header)
                            if (getSanitizedFieldFromHeaders("cache-control", resp.header) == nil) then
                                ngx.log(ngx.ERR, "missing cache control on sub request url: " .. ssiRequests[i][1])
                                totalMissingCacheControlCount = totalMissingCacheControlCount + 1
                            end
                            if respCacheControlMaxAge < minimumCacheControlMaxAge then
                                ngx.log(ngx.DEBUG, "sub request cache-control: " .. tostring(respCacheControlMaxAge))
                                ngx.ctx.ssiMinimizeMaxAgeUrl = string.sub(ssiRequests[i][1], string.len(prefix) + 1)
                                ngx.ctx.ssiMinimizeMaxAgeAge = respCacheControlMaxAge
                                ngx.ctx.ssiMinimizeMaxAgeCacheControl = getSanitizedFieldFromHeaders("cache-control", resp.header)
                                minimumCacheControlMaxAge = respCacheControlMaxAge
                            end
                        end

                        if validateJson and validateJsonInline
                        then
                            local bodyWithoutSsiIncludes = resp.body
                            for i,captureRegularExpression in ipairs(captureRegularFileExpressions) do
                                local regularExpression = string.gsub(captureRegularExpression, "([%(%)])", "")
                                bodyWithoutSsiIncludes = string.gsub(bodyWithoutSsiIncludes, regularExpression, "{}")
                            end
                            local value, errorMessage = cjson.decode(bodyWithoutSsiIncludes)
                            if (errorMessage) then
                                resp.body = generateJsonErrorFallback(string.sub(ssiRequests[i][1], string.len(prefix) + 1), errorMessage)
                                ssiResponses[ssiRequests[i][1]] = resp
                            else
                                ssiResponses[ssiRequests[i][1]] = resp
                            end
                        end

                        ssiResponses[ssiRequests[i][1]] = resp
                        -- process the response table "resp"
                    end
                end


                for i,captureRegularExpression in ipairs(captureRegularFileExpressions) do
                    local regularExpression = string.gsub(captureRegularExpression, "([%(%)])", "")

                    local replacer = function(w)
                        local ssiVirtualPath = string.match(w, captureRegularExpression)
                        if (ssiResponses[prefix .. ssiVirtualPath] == nil)
                        then
                            ngx.log(ngx.ERR, "did not capture multi with ssiVirtualPath ", ssiVirtualPath)
                            return w
                        else
                            return ssiResponses[prefix .. ssiVirtualPath].body
                        end
                    end

                    body = string.gsub(body, regularExpression, replacer)
                end

                totalSsiSubRequestsCount = totalSsiSubRequestsCount + ssiRequestsCount
                totalSsiIncludesCount = totalSsiIncludesCount + ssiMatchesCount
                ssiRequests, ssiRequestsCount, ssiMatchesCount = getSsiRequestsAndCount(ssiResponses, body)
            end
        end

        if ngx.status == 200
        then
            local md5 = ngx.md5(body)
            ngx.ctx.etag = '"' .. md5 .. '"'
        end

--        ngx.log(ngx.DEBUG, "ssiRequestsCount", totalSsiSubRequestsCount)
        ngx.ctx.ssiMissingCacheControlCount = totalMissingCacheControlCount
        ngx.ctx.ssiRequestsCount = totalSsiSubRequestsCount
        ngx.ctx.ssiIncludesCount = totalSsiIncludesCount
        ngx.ctx.ssiDepth = totalSsiDepth

        if cjson and validateJson
        then
            local value, errorMessage = cjson.decode(body)
            if errorMessage then
                body = generateJsonErrorFallback(ngx.var.request_uri, errorMessage)

                if totalSsiSubRequestsCount ~= 0 and not validateJsonInline
                then
                    local bodyTable = cjson.decode(body)
                    bodyTable.brokenSsiRequests = {}
                    -- loop over the responses table
                    for ssiRequestUrl, ssiResponse in pairs(ssiResponses) do
                        local ssiResponseBody = ssiResponse.body
                        for i,captureRegularExpression in ipairs(captureRegularFileExpressions) do
                            local regularExpression = string.gsub(captureRegularExpression, "([%(%)])", "")
                            ssiResponseBody = string.gsub(ssiResponseBody, regularExpression, "{}")
                        end
                        local ssiResponseDecodedValue, ssiResponseDecodingErrorMessage = cjson.decode(ssiResponseBody)
                        if (ssiResponseDecodingErrorMessage)
                        then
                            table.insert(bodyTable.brokenSsiRequests, {url = string.sub(ssiRequestUrl, string.len(prefix) + 1), message = ssiResponseDecodingErrorMessage })
                        end
                    end

                    body = cjson.encode(bodyTable)
                end

                ngx.ctx.etag = nil
                ngx.ctx.overrideContentType = "application/json";
                ngx.status = 500
            end
        end

    end

   if minimizeMaxAge then
       if minimumCacheControlMaxAge > 0
       then
           ngx.ctx.overrideCacheControl = "max-age=" .. minimumCacheControlMaxAge;

           if rootCacheControlSwr and minimizeOverrideStaleWhileRevalidate then
               ngx.ctx.overrideCacheControl = ngx.ctx.overrideCacheControl .. ", stale-while-revalidate=" .. minimizeOverrideStaleWhileRevalidate
           end
       else
           ngx.ctx.overrideCacheControl = "no-cache, max-age=0";
       end

   end

    ngx.ctx.res = res
    ngx.print(body)
end
